apps/lib/apps.c

```diff
@@ -1982,19 +1976,61 @@ static int do_sign_init(EVP_MD_CTX *ctx, EVP_PKEY *pkey,
         && do_pkey_ctx_init(pkctx, sigopts);
 }
 
-/* Ensure RFC 5280 compliance and then sign the certificate info */
+static int adapt_KID(X509 *cert, const STACK_OF(X509_EXTENSION) *exts,
+                     X509V3_CTX *ext_ctx, const char *name, const char *value,
+                     int add_default)
+{
+    X509_EXTENSION *new_ext = X509V3_EXT_nconf(NULL, ext_ctx, name, value);
+    int idx, rv = 0;
+
+    if (new_ext == NULL)
+        return rv;
+
+    idx = X509v3_get_ext_by_OBJ(exts, X509_EXTENSION_get_object(new_ext), -1);
+    if (idx >= 0) {
+        X509_EXTENSION *found_ext = X509v3_get_ext(exts, idx);
+        ASN1_OCTET_STRING *data = X509_EXTENSION_get_data(found_ext);
+        int KID_disabled = ASN1_STRING_length(data) <= 2; /* config said "none" */
+
+        X509_delete_ext(cert, idx);
+        X509_EXTENSION_free(found_ext);
+        rv = KID_disabled || X509_add_ext(cert, new_ext, -1);
```

Comment:
Is this right - this means the existing extension is always overridden with the default new_ext value unless completely disabled, isn't it?

Reply:
Good catch - yes, this behavior was intended, and this is because any pre-existing subject key identifier and authority key identifier might have become outdated because the included public key or its signature may have changed.
On the other hand, always overriding any existing key identifiers may not be what the user wants, in particular in case the user preferred a different style rather than what "hash" or "keyid, issuer" provides.

So I've reverted this such that any pre-existing key identifiers are retained (which is consistent with what I had written `x509v3_config.pod`). As a consequence, the user is responsible for re-generating/updating the key identifiers (by explicitly profiding specifications for them) in case needed.